integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/extension/BrokerCluster.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.test.kafkacluster.extension;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import io.kroxylicious.test.kafkacluster.KafkaCluster;
+
+/**
+ * {@code @BrokerCluster} can be used to annotate a field in a test class or a
+ *  parameter in a lifecycle method or test method of type {@link KafkaCluster}
+ *  that should be resolved into a temporary Kafka cluster.
+ *
+ * <p>A simple example looks like:</p>
+ * <pre>{@code
+ * @ExtendWith(KafkaClusterExtension.class)
+ * public class MyTest {
+ *     @Test
+ *     public void testKafkaClusterParameter(
+ *             @BrokerCluster(numBrokers = 3) KafkaCluster cluster) {
+ *         //... your test code using `cluster`
+ *     }
+ * }
+ * }</pre>
+ *
+ * <p>Alternatively annotate a {@code static} or member field:</p>
+ * <pre>{@code
+ * @ExtendWith(KafkaClusterExtension.class)
+ * public class MyTest {
+ *     @BrokerCluster(numBrokers = 3) KafkaCluster cluster;
+ *     @Test
+ *     public void testKafkaClusterParameter() {
+ *         //... your test code using `cluster`
+ *     }
+ * }
+ * }</pre>
+ *
+ * <p>By default a KRaft cluster with a single broker and colocated controller
+ * will be used using containers.
+ * However, the configuration of the cluster
+ * can be influenced with a number of other annotations:</p>
+ * <dl>
+ *     <dt>{@link KRaftCluster}</dt><dd>Allow specifying the number of KRaft controllers</dd>
+ *     <dt>{@link ZooKeeperCluster}</dt><dd>Allow specifying that a ZK-based cluster should be used</dd>
+ *     <dt>{@link SaslPlainAuth}</dt><dd>Will configure the cluster for SASL-PLAIN authentication</dd>
+ * </dl>
+ *
+ * <p>For example:</p>
+ * <pre>{@code
+ * @ExtendWith(KafkaClusterExtension.class)
+ * public class MyTest {
+ *     @Test
+ *     public void testKafkaClusterParameter(
+ *             @BrokerCluster(numBrokers = 3)
+ *             @KRaftCluster(numControllers = 3)
+ *             @SaslPlainAuth({
+ *                 @UserPassword(user="alice", password="foo"),
+ *                 @UserPassword(user="bob", password="bar")
+ *             })
```

Comment:
The annotation explosion was one concern I had when suggesting the extension originally. 

Did you experiment with something like
`@KafkaCluser(clusterConfig=[])` 

Then pass the `KafkaClusterConfig.builder()` into each clusterConfig instances i.e. 

```java
public interface ClusterConfig {

KafkaClusterConfig.builder customise(KafkaClusterConfig.builder builder);
}
```

So the BrokerCluster would implement as follows:
```java

public KafkaClusterConfig.builder customise(KafkaClusterConfig.builder builder) {                 
     return builder.brokersNum(this.numBrokers());
}
```

Reply:
I was playing more with this code on Friday and thinking about this over the weekend... One direction we could go for this whole idea would be as follows...

Support more provisioning strategies for test clusters. I.e. currently we have in-VM (which is fast, but not terribly flexible since you can only easily spin up a single version of the broker), and in-container (which is very flexible but slow due to testcontainers – even though we're using a native broker). There is scope for others: 
* For example you could download and run a native broker in a forked process which would allow testing different broker versions but still wouldn't be _as customisable_ as as using containers. E.g. you couldn't do a good job with SSL because you can't mess with hostnames/SAN certificates like you could in containers. 
* You could provision a cluster on a KaaS, or use an already-provisioned cluster on a KaaS.

The extension allows you to decouple your test code (which needs a cluster, with some configuration that might be specific to the test) from how it's provisioned. Abstractly, the annotations provide constraints on which provisioning mechanisms can be used. Within those constraints the extension is free to pick "the best" provisioning mechanism. Constraints could also be added externally to the code. This would allow you to default to testing locally using the fastest mechanism that satisfied the constraints (saving developer time), but configure a CI pipeline to test on (something closer to) the target platform (by adding extra constraints).

Looked at in this light, then, the annotation(s) are one way of expressing such constraints. Here are some constraints we might reasonably want to support:

* I want a KRaft-based/ZK-based cluster specifically
* I want at least N brokers
* I want _this particular broker version_ (or for a parameterised/templated test, _this range of broker versions_)
* I want a cluster with _this specific user_ preconfigured (using whatever SASL mechanism, I don't care)
* I want a cluster with _this specific user and password_ configured, using _this specific SASL_ mechanism.

Logically there are plenty of others too (e.g. the extension could provision things within the cluster as well as the cluster itself, such as quotas, ACLs). 

If this degree of abstraction is valuable we'd need:

1. To make cluster provisioning strategies pluggable (e.g. using `ServiceLoader`)
2. Make the constraining annotations flexible enough to cope with our needs (and other people's needs). Ideally they might even be extensible (e.g. using a meta-annotation on some custom annotation honoured by a pluggable provisioning strategy).

I'm not saying we should do this all at once and right now. But the decisions we make here could enable/prevent doing this in the future. My gut feeling is that if we did a good job with this, documented it well and promoted it then this extension could be broadly useful for many people in Red Hat and beyond. Or maybe I'm massively overengineering this. Wdyt?